### PR TITLE
Column pruning for multi-step plan

### PR DESCRIPTION
**Pull request by zengyan1**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #10096 

## What this PR does / why we need it:

### CODE REVIEW

**Comment by zengyan1** [comment]
waiting for #10860 to be merged first

**Comment by aunjgr** [comment]
should cherry-pick the top commit and force-push after #10860 is merged

**Comment by m-schen** [comment]
very similar to pr 10860...

**Review by aunjgr — APPROVED** [review]

**Pull request merged**
